src/lib.rs

```diff
@@ -308,165 +173,288 @@ impl KafkaJsonToDelta {
-                    //
-
-                    let should_complete_record_batch = json_buffer.len()
-                        == self.max_messages_per_batch
-                        || latency_timer.elapsed().as_secs() >= self.allowed_latency;
-
                     debug!(
-                        "Should complete record batch? {}, json_buffer.len: {}. timer.elapsed: {}",
-                        should_complete_record_batch,
-                        json_buffer.len(),
-                        latency_timer.elapsed().as_secs()
+                        "Latency timer at {} milliseconds.",
+                        latency_timer.elapsed().as_millis()
                     );
 
-                    if should_complete_record_batch {
-                        info!("Creating Arrow RecordBatch from JSON message buffer.");
+                    // Finalize arrow record batch
+                    if self.should_complete_record_batch(&value_buffers, &latency_timer) {
+                        info!("Creating Arrow RecordBatch.");
 
-                        // Convert the json buffer to an arrow record batch.
-                        let record_batch = deltalake_ext::record_batch_from_json_buffer(
-                            arrow_schema_ref.clone(),
-                            json_buffer.as_slice(),
-                        )?;
+                        let (values, partition_offsets) = consume_value_buffers(
+                            self.partition_assignment.clone(),
+                            &mut value_buffers,
+                        )
+                        .await?;
 
-                        // Clear the json buffer so we can start clean on the next round.
-                        json_buffer.clear();
+                        let record_batch = deltalake_ext::record_batch_from_json(
+                            delta_writer.arrow_schema(),
+                            values.as_slice(),
+                        )?;
 
-                        // Write the arrow record batch to our delta writer that wraps an in memory cursor for tracking bytes.
-                        delta_writer
-                            .write_record_batch(&partition_cols, &record_batch)
-                            .await?;
+                        // Since we are writing a record batch to the in-memory parquet file, value buffers must be marked
+                        // dirty so we know we have to drop all buffers and re-seek if a rebalance occurs before the next file write.
+                        value_buffers.mark_dirty();
 
-                        // When the memory buffer meets min bytes per file or allowed latency is met, complete the file and start a new one.
-                        let should_complete_file = delta_writer.buffer_len()
-                            >= self.min_bytes_per_file
-                            || latency_timer.elapsed().as_secs() >= self.allowed_latency;
+                        delta_writer.write_record_batch(&record_batch).await?;
```

Comment:
This `?` means that a failure to write the record batch will cause the run_loop to error up into the main function and exit the `tokio::spawn` (as far as I can tell) and exit the process.

Is that the desired behavior? There's a lot of potential errors that look like they would be hiding behind that `write_record_batch` call